pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java

```diff
@@ -0,0 +1,205 @@
+    return aggregateTableMetadataInfo;
+  }
+
+  /**
+   * This method is called when the API request is to fetch segment metadata for all segments of the table.
+   * This method makes a MultiGet call to all servers that host their respective segments and gets the results.
+   * This method accept a list of column names as filter, and will return column metadata for the column in the
+   * list.
+   * @return list of segments and their metadata as a JSON string
+   */
+  public List<String> getSegmentMetadataFromServer(String tableNameWithType,
+      Map<String, List<String>> serversToSegmentsMap, BiMap<String, String> endpoints, List<String> columns,
+      int timeoutMs) {
+    LOGGER.debug("Reading segment metadata from servers for table {}.", tableNameWithType);
+    List<String> serverURLs = new ArrayList<>();
+    for (Map.Entry<String, List<String>> serverToSegments : serversToSegmentsMap.entrySet()) {
+      List<String> segments = serverToSegments.getValue();
+      for (String segment : segments) {
+        serverURLs.add(generateSegmentMetadataServerURL(tableNameWithType, segment, columns,
+            endpoints.get(serverToSegments.getKey())));
+      }
+    }
+    BiMap<String, String> endpointsToServers = endpoints.inverse();
+    CompletionServiceHelper completionServiceHelper =
+        new CompletionServiceHelper(_executor, _connectionManager, endpointsToServers);
+    CompletionServiceHelper.CompletionServiceResponse serviceResponse =
+        completionServiceHelper.doMultiGetRequest(serverURLs, tableNameWithType, true, timeoutMs);
+    List<String> segmentsMetadata = new ArrayList<>();
+
+    int failedParses = 0;
+    for (Map.Entry<String, String> streamResponse : serviceResponse._httpResponses.entrySet()) {
+      try {
+        String segmentMetadata = streamResponse.getValue();
+        segmentsMetadata.add(segmentMetadata);
+      } catch (Exception e) {
+        failedParses++;
+        LOGGER.error("Unable to parse server {} response due to an error: ", streamResponse.getKey(), e);
+      }
+    }
+    if (failedParses != 0) {
+      LOGGER.error("Unable to parse server {} / {} response due to an error: ", failedParses, serverURLs.size());
+    }
+
+    LOGGER.debug("Retrieved segment metadata from servers.");
+    return segmentsMetadata;
+  }
+
+  private String generateAggregateSegmentMetadataServerURL(String tableNameWithType, List<String> columns,
+      String endpoint) {
+    try {
+      tableNameWithType = URLEncoder.encode(tableNameWithType, StandardCharsets.UTF_8.name());
+      String paramsStr = "";
+      if (columns != null) {
+        List<String> params = new ArrayList<>(columns.size());
+        for (String column : columns) {
+          params.add(String.format("columns=%s", column));
+        }
+        paramsStr = String.join("&", params);
+      }
+      return String.format("%s/tables/%s/metadata?%s", endpoint, tableNameWithType, paramsStr);
+    } catch (UnsupportedEncodingException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)